### PR TITLE
Render loop

### PR DESCRIPTION
**Pull request by grokys**
This PR adds a render loop. It's intended to be reasonably non-intrusive, and the beginning of a larger overhaul of our rendering.
- It adds the start of a `RenderTest` sample which we can use to monitor performance/diagnose drawing errors etc
- Adds an `IRenderLoop` interface, a default `DefaultRenderLoop` implementation and a `Win32.RenderLoop` implementation that uses multimedia timers (the default Win32 timer only has the resolution to go up to about 30 fps when requested 60fps)
- It replaces `IRenderRoot.RenderQueueManager` with `IRenderRoot.Renderer` which is of type `IRenderer`
- `IRenderer` will keep track of dirty rectangles (currently unused) and carry out rendering. For rendering it currently just calls the existing rendering infrastructure
- It adds `IRenderRoot.CreateRenderTarget` to allow a render root to create a render target (this currently just calls `IPlatformRenderInterface.CreateRenderer`) as before, but it is intended to decouple the renderer from knowledge of platform handles (which are only defined higher up in the stack)

### CODE REVIEW

**Comment by grokys** [comment]
cc: @kekekeks

File: src/Avalonia.SceneGraph/Rendering/Renderer.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    public class Renderer : IDisposable, IRenderer
+    {
+        private IRenderLoop _renderLoop;
+        private IRenderRoot _root;
+        private IRenderTarget _renderTarget;
+        private bool _dirty;
+
+        public void Attach(IRenderRoot root, IRenderLoop renderLoop)
+        {
+            Contract.Requires<ArgumentNullException>(root != null);
```
Comment (kekekeks) [inline src/Avalonia.SceneGraph/Rendering/Renderer.cs]: Is renderer instance reusable? If it's not why not to use constructor injection or at least check that `Attach` was already called?

File: src/Avalonia.SceneGraph/Rendering/Renderer.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+            try
+            {
+                _renderTarget.Render(_root);
+            }
+            catch (RenderTargetCorruptedException ex)
+            {
+                Logging.Logger.Information("Renderer", this, "Render target was corrupted. Exception: {0}", ex);
+                _renderTarget.Dispose();
+                _renderTarget = null;
```
Comment (kekekeks) [inline src/Avalonia.SceneGraph/Rendering/Renderer.cs]: I think we are supposed to retry render pass here

File: src/Windows/Avalonia.Win32/RenderLoop.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        public RenderLoop(int framesPerSecond)
+            : base(framesPerSecond)
+        {
+        }
+
+        protected override IDisposable StartCore(Action tick)
+        {
+            timerDelegate = (id, uMsg, user, dw1, dw2) => tick();
+
+            var handle = UnmanagedMethods.timeSetEvent(
```
Comment (kekekeks) [inline src/Windows/Avalonia.Win32/RenderLoop.cs]: Are you sure that it's thread safe? MSDN states that
>The multimedia timer runs in its own thread

File: src/Avalonia.Controls/TopLevel.cs [hunk 4]
```diff
@@ -232,6 +235,18 @@ protected bool AutoSizing
         }
 
         /// <inheritdoc/>
+        IRenderTarget IRenderRoot.CreateRenderTarget()
+        {
+            return _renderInterface.CreateRenderer(PlatformImpl.Handle);
```
Comment (kekekeks) [inline src/Avalonia.Controls/TopLevel.cs]: May be we should move `CreateRenderer` to `ITopLevelImpl`. Windowing platform has more information for negotiations with rendering platform than we do in `TopLevel` class

Comment (kekekeks) [inline src/Avalonia.Controls/TopLevel.cs]: I think we might even make `ITopLevelImpl` to actually implement `IRenderTarget` by itself. There are some cases (e. g. GTK embedding) when we are trying to obtain render target when native window haven't been created yet. I think we should ask for `DrawingContext` from `ITopLevelImpl` directly and allow it to return `null` if `DrawingContext` can't be created and skip render pass in this case. It will also allow to skip frames when window isn't effectively visible to the user (minimized, out of screen borders, etc).

File: src/Avalonia.SceneGraph/Rendering/Renderer.cs (hunk 1 above)
Comment (grokys) [inline src/Avalonia.SceneGraph/Rendering/Renderer.cs]: Nope, there is one renderer per top level control. I use `Attach` because it's created with the service locator and there's no way to pass constructor arguments using that.

File: src/Avalonia.SceneGraph/Rendering/Renderer.cs (hunk 2 above)
Comment (grokys) [inline src/Avalonia.SceneGraph/Rendering/Renderer.cs]: Well the next render pass will happen in 1/60 of a second, so I figured just wait until then.

File: src/Windows/Avalonia.Win32/RenderLoop.cs (hunk 3 above)
Comment (grokys) [inline src/Windows/Avalonia.Win32/RenderLoop.cs]: Yep, it runs in its own thread, it's the responsibility of whatever subscribes to the render loop to handle passing the notification to the correct thread - I should document that fact.

At the moment we're still going via WM_PAINT to do the rendering so that will move us onto the right thread; later on we'll probably paint on the render thread to an offscreen bitmap and render sections of that on WM_PAINT as needed.

File: src/Avalonia.Controls/TopLevel.cs (hunk 4 above)
Comment (grokys) [inline src/Avalonia.Controls/TopLevel.cs]: Right, that's what I was going for with this. `IRenderRoot` is implemented by `TopLevel` and that can choose how to handle it; including passing the request to `ITopLevelImpl`. 

If I understand you correctly you're talking about losing `IRenderTarget` too, and just having something like `IRenderRoot.CreateDrawingContext` (which TopLevel will choose to pass to `ITopLevelImpl.CreateRenderTarget`)? If so, I think that's a good idea.

File: src/Avalonia.SceneGraph/Rendering/Renderer.cs (hunk 1 above)
Comment (kekekeks) [inline src/Avalonia.SceneGraph/Rendering/Renderer.cs]: You could register `Func<TopLevel, IRenderer>` (or custom delegate to be used as factory method) in locator instead

File: src/Avalonia.Controls/TopLevel.cs (hunk 4 above)
Comment (kekekeks) [inline src/Avalonia.Controls/TopLevel.cs]: Yep, something like that. IRenderTarget isn't that useful right now anyway, since it provides just one method.

Comment (kekekeks) [inline src/Avalonia.Controls/TopLevel.cs]: Well, `IRenderTarget` *is* useful for tracking state from the rendering system (direct2d/opengl context, offscreen surface, etc), but lifetime management of said render target should be the job of `ITopLevelImpl`, I think.

File: src/Avalonia.SceneGraph/Rendering/Renderer.cs (hunk 1 above)
Comment (grokys) [inline src/Avalonia.SceneGraph/Rendering/Renderer.cs]: Yeah, true. Or even register a `RendererFactory`...

File: src/Avalonia.Controls/TopLevel.cs (hunk 4 above)
Comment (kekekeks) [inline src/Avalonia.Controls/TopLevel.cs]: I think `ITopLevelImpl` implementation should obtain `IRenderTarget` from rendering subsystem (using IPlatformHandle *or* bitmap-backed) and manage it's lifetime by itself.

Comment (grokys) [inline src/Avalonia.Controls/TopLevel.cs]: Ok, sounds good - though I think that should be a separate PR; I'd like to keep this focused on the render loop change. As I said it's intended to be part of a much larger change. As such does it look good to merge now to you? If so could you merge? I'd like to use the system where the reviewer merges the PR rather than the author.

File: src/Avalonia.SceneGraph/Rendering/Renderer.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            {
+                _dirty = false;
+            }
+        }
+
+        private void OnRenderLoopTick(object sender, EventArgs e)
+        {
+            if (_dirty)
+            {
+                _root.Invalidate(new Rect(_root.ClientSize));
```
Comment (kekekeks) [inline src/Avalonia.SceneGraph/Rendering/Renderer.cs]: That's being called on `IRenderLoop` thread, so it might not be safe to call `IRenderRoot` methods here. For now on win32 platform `Invalidate` just calls `ITopLevel::Invalidate` which in turn calls `SendMessage`, so you've got away with that, but it is not safe in general case.

I also propose to add some suffix for events that might trigger on non-ui thread (e. g. `Tick` should be named `TickThreadUnsafe`) to make it easier to track such things.

Comment (grokys) [inline src/Avalonia.SceneGraph/Rendering/Renderer.cs]: Yep, this is done with the knowledge that we go via the message loop for paints. It will be replaced by the real render code in upcoming patches.

Comment (grokys) [inline src/Avalonia.SceneGraph/Rendering/Renderer.cs]: I guess synchronizing with the UI thread here wouldn't be a problem though.

**Comment by grokys** [comment]
@kekekeks given that the remaining comments you raised will be fixed in upcoming PRs, are we OK to merge this?
